### PR TITLE
[KS-420] Add rate-limiting to Dispatcher

### PR DESCRIPTION
**Pull request by bolekk**

### CODE REVIEW

File: core/config/toml/types.go [hunk 1]
```diff
@@ -1434,8 +1434,49 @@ func (r *ExternalRegistry) setFrom(f *ExternalRegistry) {
...
+
+	if f.SupportedVersion != nil {
+		d.SupportedVersion = f.SupportedVersion
+	}
+}
+
+type DispatcherRateLimit struct {
+	GlobalRPS   *float64
+	GlobalBurst *int
+	RPS         *float64
```
Comment (bolekk) [inline core/config/toml/types.go]: nit: Maybe call them PerSenderRPS / PerSenderBurst to be super clear?

File: core/capabilities/remote/dispatcher.go [hunk 2]
```diff
@@ -168,6 +179,11 @@ func (d *dispatcher) receive() {
 			d.lggr.Info("stopped - exiting receive")
 			return
 		case msg := <-recvCh:
+			if !d.rateLimiter.Allow(msg.Sender.String()) {
+				d.lggr.Debugw("rate limit exceeded, dropping message", "sender", msg.Sender)
+				// TODO: do we just drop the message, or do we need to respond with an error?
```
Comment (bolekk) [inline core/capabilities/remote/dispatcher.go]: Just drop. We don't want to respond as this is a protection against spammy nodes.

File: core/config/docs/core.toml [hunk 3]
```diff
@@ -450,6 +450,22 @@ NetworkID = 'evm' # Default
...
+
+[Capabilities.Dispatcher.RateLimit]
+# GlobalRPS is the global rate limit for the dispatcher.
+GlobalRPS = 800 # Default
+# GlobalBurst is the global burst limit for the dispatcher.
+GlobalBurst = 1000 # Default
+# PerSenderRPS is the per-sender rate limit for the dispatcher.
+PerSenderRPS = 5 # Example
+# PerSenderBurst is the per-sender burst limit for the dispatcher.
+PerSenderBurst = 10 # Example
```
Comment (bolekk) [inline core/config/docs/core.toml]: Let's make these slightly higher: PerSenderRPS = 10, Burst = 50.

File: docs/CONFIG.md [hunk 4]
```diff
@@ -1233,6 +1233,60 @@ ChainID = '1' # Default
...
+SupportedVersion = 1 # Default
+ReceiverBufferSize = 10000 # Default
+```
+
+
+### SupportedVersion
+```toml
+SupportedVersion = 1 # Default
+```
+SupportedVersion is the version of the dispatcher that the node supports.
```
Comment (bolekk) [inline docs/CONFIG.md]: Not version of the dispatcher but version of message schema.

**Review by bolekk — COMMENTED** [review]
Overall looks good but some tests are still failing.

File: core/config/dispatcher_config.go [hunk 5]
```diff
@@ -0,0 +1,14 @@
+package config
+
+type DispatcherRateLimit interface {
+	GlobalRPS() float64
+	GlobalBurst() int
+	RPS() float64
+	Burst() int
```
Comment (bolekk) [inline core/config/dispatcher_config.go]: Can we rename these two to PerSenderRPS() and PerSenderBurst() please?

File: main_test.go [hunk 6]
```diff
@@ -56,7 +56,7 @@ func TestScripts(t *testing.T) {
 				Dir:             path,
 				Setup:           commonEnv,
 				ContinueOnError: true,
-				// UpdateScripts:   true, // uncomment to update golden files
```
Comment (bolekk) [inline main_test.go]: not needed any more

Comment (vyzaldysanchez) [reply main_test.go]: Committed by mistake.

**Review by vyzaldysanchez — APPROVED** [review]

**Review by cedric-cordenier — APPROVED** [review]

**Pull request merged**